direction_matrix.go

```diff
@@ -0,0 +1,59 @@
+package mapbox
+
+import (
+	"context"
+	"fmt"
+	"net/url"
+	"strings"
+)
+
+const (
+	directionsMatrix = "directions-matrix"
+)
+
+type DirectionsMatrixRequest struct {
+	// required
+	Profile Profile
+	Coordinates Coordinates
+
+	// optional
+	Annotations Annotations
+	Approaches Approaches
+	Destinations Destinations
+	Sources Sources
+	FallbackSpeed FallbackSpeed
+
+}
+
+type DirectionsMatrixResponse struct {
+	Code string `json:"code"`
+	Durations [][]float64 `json:"durations"`
+	Distances [][]float64 `json:"distances"`
+	Destinations []Waypoint `json:"destinations"`
+	Sources []Waypoint `json:"sources"`
+}
+
+// https://docs.mapbox.com/api/navigation/#matrix
+func (c *Client) DirectionsMatrix(ctx context.Context, req DirectionsMatrixRequest) (*DirectionsMatrixResponse, error) {
+	relPath := fmt.Sprintf("%v/%v/%v/%v", directionsMatrix, v1, req.Profile, req.Coordinates.WGS84Format())
+
+	query := url.Values{}
+	query.Set("access_token", c.apiKey)
+	query.Set("annotations", strings.Join(req.Annotations.strings(), ","))
```

Comment:
Are things like annotations, etc used for other queries besides `DirectionsMatrix`?

If so we should move this join statement onto methods in `types.go`.

Either rename `strings()` to `string()` or have both if we need both. Or `asParam()` or something might also be a good name if you don't like the string names.

That way every endpoint doesn't have a bunch of strings.Join() calls that have to know if they join with semicolon or comma, etc. And every endpoint that we add will just call .string() on it.

Reply:
good callout